app/src/main/java/com/muzima/service/MuzimaLoggerService.java

```diff
@@ -0,0 +1,167 @@
+import com.muzima.api.context.Context;
+import com.muzima.api.model.MuzimaSetting;
+import com.muzima.api.model.User;
+import com.muzima.controller.MuzimaSettingController;
+import com.muzima.model.location.MuzimaGPSLocation;
+import com.muzima.util.MuzimaLogger;
+import net.minidev.json.JSONObject;
+import org.json.JSONException;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Timer;
+import java.util.UUID;
+
+import static com.muzima.util.Constants.ServerSettings.LOGGING_FEATURE_ENABLED_SETTING;
+
+public class MuzimaLoggerService {
+    private static String pseudoDeviceId = null;
+
+    private static Timer timer;
+
+    public static String getAndParseGpsLocationForLogging(final MuzimaApplication muzimaApplication){
+        if(isLoggingFeatureEnabled(muzimaApplication)) {
+            MuzimaGPSLocationService muzimaLocationService = muzimaApplication.getMuzimaGPSLocationService();
+
+            HashMap<String, Object> locationDataHashMap = muzimaLocationService.getLastKnownGPS();
+            if (locationDataHashMap.containsKey("gps_location")) {
+                MuzimaGPSLocation muzimaGPSLocation = ((MuzimaGPSLocation) locationDataHashMap.get("gps_location"));
+                try {
+                    return muzimaGPSLocation.toJsonObject().toString();
+                } catch (JSONException e) {
+                    Log.e("MuzimaLoggerService", "Error while obtaining GPS location", e);
+                }
+            } else if (locationDataHashMap.containsKey("gps_location_status")) {
+                return "{\"gps_location_status\":\"" + locationDataHashMap.get("gps_location_status") + "\"}";
+            }
+        }
+        return "{}";
+    }
+
+    public static void log(final MuzimaApplication muzimaApplication, final String tag, final String userId, final String gpsLocation, final String details){
+        if(isLoggingFeatureEnabled(muzimaApplication)) {
+
+            new AsyncTask<Void, Void, Void>() {
+                protected Void doInBackground(Void... voids) {
+                    String deviceId = getPseudoDeviceId();
+                    JSONObject timestamp = new JSONObject(){{
+                        put("sntpTimestamp",muzimaApplication.getSntpService().getLocalTime(1000).getTime());
+                        put("systemTimestamp",System.currentTimeMillis());
+                    }};
+
+                    MuzimaLogger.log(muzimaApplication.getMuzimaContext(), tag, userId, gpsLocation, details, deviceId,timestamp.toJSONString());                    return null;
+                }
+            }.execute();
+        }
+    }
+
+    public static void log(final MuzimaApplication muzimaApplication, final String tag, final String details){
+        if(isLoggingFeatureEnabled(muzimaApplication)) {
+            User authenticatedUser = muzimaApplication.getAuthenticatedUser();
+            if (authenticatedUser != null) {
+                String userId = authenticatedUser.getUuid();
```

Comment:
I think this is fine, because the parameter is a userId, and the uuid is just a representation of it. Other places that cannot access the uuid use username (e.g login activity). In the future, this may be changed to something else